### ISSUE TITLE
False positive "spawned process is never waited on" when using BufReader(cmd.stdout)

### ISSUE THREAD

**Issue by th1000s** [rust-lang/rust-clippy#13748]
### Summary

Clippy false positive, `strace -e process  ..` even shows the `wait4()` syscall.

### Lint Name

clippy::zombie_processes

### Reproducer

I tried this code:

```rust
use std::io::BufRead;
use std::process::{Command, Stdio};

fn main() {
    cmd();
}

pub fn delta<I>(_lines: I) -> std::io::Result<()>
where
    I: BufRead,
{
    Ok(())
}

fn cmd() -> Option<i32> {
    let cmd = Command::new("true").stdout(Stdio::piped()).spawn();

    let mut cmd = cmd.unwrap();

    let cmd_stdout = cmd.stdout.as_mut().unwrap();
    let cmd_stdout_buf = std::io::BufReader::new(cmd_stdout);

    let res = delta(cmd_stdout_buf);

    // Remove this block and the warning goes away:
    if let Err(error) = res {
        match error.kind() {
            std::io::ErrorKind::BrokenPipe => return Some(0),
            _ => {
                return Some(1);
            }
        }
    };

    Some(cmd.wait().unwrap().code().unwrap()) // <-- wait()!
}

```

I saw this happen:

```
warning: spawned process is never `wait()`ed on
  --> src/main.rs:18:19
   |
18 |     let mut cmd = cmd.unwrap();
   |                   ^^^^^^^^^^^^
   |
   = note: consider calling `.wait()`
   = note: not doing so might leave behind zombie processes
   = note: see https://doc.rust-lang.org/stable/std/process/struct.Child.html#warning
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#zombie_processes
   = note: `#[warn(clippy::zombie_processes)]` on by default

warning: `s` (bin "s") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.05s
```

(see [Rust Playground ](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=95b46720a9c83bb37522cbbe363620a2))

I expected to see this happen:

No warning.

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.1
```

**Comment by Jarcho** [comment]
You're early return branch isn't waited on which is what the lint is supposed to detect. The error message itself is definitely wrong though.

**Comment by th1000s** [comment]
Right, but still _not_ a false positive. Thank you. It might mention that it is not waited for in all code paths. Updated.

**Comment by y21** [comment]
I was working on this lint anyway, so: @rustbot claim
